beacon-chain/cache/sync_committee.go

```diff
@@ -0,0 +1,187 @@
+	defer s.lock.RUnlock()
+
+	pos, err := s.idxPositionInCommittee(root, valIdx)
+	if err != nil {
+		return nil, err
+	}
+	if pos == nil {
+		return []types.CommitteeIndex{}, nil
+	}
+
+	return pos.currentPeriod, nil
+}
+
+// NextPeriodIndexPosition returns next period index position of a validator index in respect with sync committee.
+// If the input validator index has no assignment, an empty list will be returned.
+// If the input root does not exist in cache, `ErrNonExistingSyncCommitteeKey` is returned.
+// Manual checking of state for index position in state is recommended when `ErrNonExistingSyncCommitteeKey` is returned.
+func (s *SyncCommitteeCache) NextPeriodIndexPosition(root [32]byte, valIdx types.ValidatorIndex) ([]types.CommitteeIndex, error) {
+	s.lock.RLock()
+	defer s.lock.RUnlock()
+
+	pos, err := s.idxPositionInCommittee(root, valIdx)
+	if err != nil {
+		return nil, err
+	}
+	if pos == nil {
+		return []types.CommitteeIndex{}, nil
+	}
+	return pos.nextPeriod, nil
+}
+
+// Helper function for `CurrentPeriodIndexPosition` and `NextPeriodIndexPosition` to return a mapping
+// of validator index to its index(s) position in the sync committee.
+func (s *SyncCommitteeCache) idxPositionInCommittee(
+	root [32]byte, valIdx types.ValidatorIndex,
+) (*positionInCommittee, error) {
+	obj, exists, err := s.cache.GetByKey(key(root))
+	if err != nil {
+		return nil, err
+	}
+	if !exists {
+		SyncCommitteeCacheMiss.Inc()
+		return nil, ErrNonExistingSyncCommitteeKey
+	}
+	item, ok := obj.(*syncCommitteeIndexPosition)
+	if !ok {
+		return nil, errNotSyncCommitteeIndexPosition
+	}
+	idxInCommittee, ok := item.vIndexToPositionMap[valIdx]
+	if !ok {
+		SyncCommitteeCacheMiss.Inc()
+		return nil, nil
+	}
+	SyncCommitteeCacheHit.Inc()
+	return idxInCommittee, nil
+}
+
+// UpdatePositionsInCommittee updates caching of validators position in sync committee in respect to
+// current epoch and next epoch. This should be called when `current_sync_committee` and `next_sync_committee`
+// change and that happens every `EPOCHS_PER_SYNC_COMMITTEE_PERIOD`.
+func (s *SyncCommitteeCache) UpdatePositionsInCommittee(syncCommitteeBoundaryRoot [32]byte, st state.BeaconStateAltair) error {
+	csc, err := st.CurrentSyncCommittee()
+	if err != nil {
+		return err
+	}
+	positionsMap := make(map[types.ValidatorIndex]*positionInCommittee)
+	for i, pubkey := range csc.Pubkeys {
+		p := bytesutil.ToBytes48(pubkey)
+		validatorIndex, ok := st.ValidatorIndexByPubkey(p)
+		if !ok {
+			continue
+		}
+		if _, ok := positionsMap[validatorIndex]; !ok {
+			m := &positionInCommittee{currentPeriod: []types.CommitteeIndex{types.CommitteeIndex(i)}, nextPeriod: []types.CommitteeIndex{}}
+			positionsMap[validatorIndex] = m
+		} else {
+			positionsMap[validatorIndex].currentPeriod = append(positionsMap[validatorIndex].currentPeriod, types.CommitteeIndex(i))
+		}
+	}
+
+	nsc, err := st.NextSyncCommittee()
+	if err != nil {
+		return err
+	}
+	for i, pubkey := range nsc.Pubkeys {
```

Comment:
Some duplication in these two for loops that could be refactored perhaps

Reply:
Can't refactor this, one is current period and one is next period. The structure is very different

Reply:
it's doable, but it would be a weird abstraction you would need to do with callbacks..not worth it